### DISCUSSION TITLE
TSML "Import Data Sources" duplicating fields

### DISCUSSION THREAD

**Discussion by sblack4** [code4recovery/12-step-meeting-list#1818] [Q&A]
** The images in here have failed to copy paste. See https://docs.google.com/document/d/1DgWuN0Uo8aPzW9hxyYMcr8PiJlLn201TZKNr9xpZMUQ/edit?usp=sharing 

# What happened on Friday when we were re-importing the feeds

This is what it was like

![][image1]

But then when we imported a CSV of our meeting list, minus two districts we are now going to import with “delete all meetings, locations, and groups not from a data source” selected and got this. We tried “delete all meetings, locations, groups, districts, and regions” but got the same thing except the old Blueridge data sources showed 0 meetings.   
![][image2]

# Weird behavior 2025-11-02

Plugin Version 3.19.7 

If I click “Refresh” on one of the upper or lower  [aablueridge.org](http://aablueridge.org) sources I get “Your meeting list is already in sync with the feed.’

![][image3]

If I try to delete one of the upper or lower  [aablueridge.org](http://aablueridge.org) sources it deletes the lower one. I clicked “X” on the first one:  
![][image4]

I can delete the newer, lower ones. I clicked “X” on the top of the bottom group and it deleted:   
![][image5]

Now if I try refreshing one of the top ones (I tried refreshing the top one) It successfully refreshes: 

![][image6]

And for the kicker: we can’t delete the upper data sources. I clicked “X’ next to the second of the upper group and OK to “Remove the aablueridge.org feed?” but when the page reloads it’s still there:  
![][image7]

The four data sources in question do come from the same host:

- ​​[https://aablueridge.org//wp-admin/admin-ajax.php?district=district-26-winchester\&district-35-fauquier\&district-39-piedmont\&district-40-skyline\&district-unknown\&action=meetings](https://aablueridge.org//wp-admin/admin-ajax.php?district=district-26-winchester&district-35-fauquier&district-39-piedmont&district-40-skyline&district-unknown&action=meetings)   
- [https://aablueridge.org//wp-admin/admin-ajax.php?district=district-35-fauquier\&district-39-piedmont\&district-40-skyline\&district-unknown\&action=meetings](https://aablueridge.org//wp-admin/admin-ajax.php?district=district-35-fauquier&district-39-piedmont&district-40-skyline&district-unknown&action=meetings)   
- [https://aablueridge.org/wp-admin/admin-ajax.php?district=district-39-piedmont\&district-40-skyline\&district-unknown\&action=meetings](https://aablueridge.org/wp-admin/admin-ajax.php?district=district-39-piedmont&district-40-skyline&district-unknown&action=meetings)   
- [https://aablueridge.org/wp-admin/admin-ajax.php?district=district-40-skyline\&district-unknown\&action=meetings](https://aablueridge.org/wp-admin/admin-ajax.php?district=district-40-skyline&district-unknown&action=meetings)

Any ideas as to what is causing these issues? 
Thank you!!!

**Comment by SIA-WEBSUP** [comment]
@sblack4 ,

We're trying to repo this, with no luck so far.
- I was able to add  all the sources you listed to my test site and our demo site with no issues
- if I attempted to add the same Data Source (i.e. identical URL) but with a different name, I get a "Your meeting list is already in sync with the feed." notification and the second instance of the feed is not added
- I was able to delete the feeds with no issues

So I'm not sure how you were able to add a second entry for an identical source. Or was the source slightly different (i.e. some but not all of the districts for a source that was already there?)

I also noticed that two of the four URLs you posted had some extra forward slashes near the beginning, but I don't think this matters at all to either the browser or the plugin. Just an observation.

If you're able to, I would suggest deleting all the feeds and starting over. But I think you're saying that you aren't able to do so?
- do you have the flexibility to attempt to delete all the external data sources?
- are you ***able*** to do so?
- tells us again what happens  when you try

**Reply by sblack4** [reply]
Ok, so I ran [normalize-data-sources.php](https://gist.github.com/sblack4/95318fa5b55651fa323250f92fd8d2f6) to fix some encoding issues
The duplicates were caused by URLs stored with inconsistent formats:
- HTML entities (\&amp; vs &)
- Double slashes (// vs /)
- Different URL encodings
Solution:
Created a one-time normalization script that:
1. Normalizes all data source URLs (decodes HTML entities, removes double slashes)
2. Identifies and merges duplicates (keeps the entry with the most recent last_import timestamp)
3. Updates meeting post meta to reference normalized URLs
4. Saves the cleaned-up data sources array
To fix existing duplicates:
```
wp eval-file normalize-data-sources.php
```

This reduced my data sources from 15 to 10 by removing 5 duplicates. The script is safe to run multiple times and shows what changes it makes.


before

```
WP Engine:13306/wp_aavirginia/vvx_options/		https://phpmyadmin.wpengine.com/index.php?route=/table/sql&db=wp_aavirginia&table=vvx_options

   Showing rows 0 -  0 (1 total, Query took 0.0194 seconds.)


SELECT option_value 
FROM wp_aavirginia.vvx_options 
WHERE option_name = 'tsml_data_sources';



a:15:{s:61:"http://vadist15aa.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716849;s:14:"count_meetings";i:58;s:4:"name";s:11:"District 15";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:7:"enabled";s:4:"type";s:4:"JSON";}s:66:"https://harrisonburgaa.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716861;s:14:"count_meetings";i:34;s:4:"name";s:26:"District 22 (Harrisonburg)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:7:"enabled";s:4:"type";s:4:"JSON";}s:63:"https://aaheartofva.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716929;s:14:"count_meetings";i:133;s:4:"name";s:31:"District 09 (Heart of Virginia)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:7:"enabled";s:4:"type";s:4:"JSON";}s:168:"https://aablueridge.org//wp-admin/admin-ajax.php?district=district-35-fauquier&amp;district-39-piedmont&amp;district-40-skyline&amp;district-unknown&amp;action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1761937015;s:14:"count_meetings";i:49;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";N;s:13:"change_detect";N;s:4:"type";s:4:"JSON";}s:117:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-40-skyline&amp;district-unknown&amp;action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1761937015;s:14:"count_meetings";i:51;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";N;s:13:"change_detect";N;s:4:"type";s:4:"JSON";}s:195:"https://aablueridge.org//wp-admin/admin-ajax.php?district=district-26-winchester&amp;district-35-fauquier&amp;district-39-piedmont&amp;district-40-skyline&amp;district-unknown&amp;action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1761937015;s:14:"count_meetings";i:62;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";N;s:13:"change_detect";N;s:4:"type";s:4:"JSON";}s:142:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-39-piedmont&amp;district-40-skyline&amp;district-unknown&amp;action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1761937015;s:14:"count_meetings";i:25;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";N;s:13:"change_detect";N;s:4:"type";s:4:"JSON";}s:73:"https://www.aadistrict30va.org/wp/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716962;s:14:"count_meetings";i:93;s:4:"name";s:27:"District 30 (Fredricksburg)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:90:"https://sheets.code4recovery.org/storage/1LzVrSE-t-iYMnSmiZtg8449rctjJ8Xn9l-nz9_t4H6k.json";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717027;s:14:"count_meetings";i:64;s:4:"name";s:31:"District 38 (Historic Virginia)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:75:"https://www.tidewaterintergroup.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717037;s:14:"count_meetings";i:356;s:4:"name";s:37:"Tidewater Intergroup (14, 23, 24, 32)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:61:"https://aaroanoke.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717051;s:14:"count_meetings";i:98;s:4:"name";s:21:"Roanoke (District 19)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:109:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717062;s:14:"count_meetings";i:51;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:175:"https://aablueridge.org//wp-admin/admin-ajax.php?district=district-26-winchester&district-35-fauquier&district-39-piedmont&district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717071;s:14:"count_meetings";i:61;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:130:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-39-piedmont&district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717080;s:14:"count_meetings";i:25;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:152:"https://aablueridge.org//wp-admin/admin-ajax.php?district=district-35-fauquier&district-39-piedmont&district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717089;s:14:"count_meetings";i:48;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}}	
```


after:
```
WP Engine:13306/		https://phpmyadmin.wpengine.com/index.php?route=/server/sql

   Showing rows 0 -  0 (1 total, Query took 0.0192 seconds.)


SELECT option_value 
FROM wp_aavirginia.vvx_options 
WHERE option_name = 'tsml_data_sources';



a:11:{s:61:"http://vadist15aa.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716849;s:14:"count_meetings";i:58;s:4:"name";s:11:"District 15";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:7:"enabled";s:4:"type";s:4:"JSON";}s:66:"https://harrisonburgaa.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716861;s:14:"count_meetings";i:34;s:4:"name";s:26:"District 22 (Harrisonburg)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:7:"enabled";s:4:"type";s:4:"JSON";}s:63:"https://aaheartofva.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716929;s:14:"count_meetings";i:133;s:4:"name";s:31:"District 09 (Heart of Virginia)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:7:"enabled";s:4:"type";s:4:"JSON";}s:151:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-35-fauquier&district-39-piedmont&district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717089;s:14:"count_meetings";i:48;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:109:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717062;s:14:"count_meetings";i:51;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:174:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-26-winchester&district-35-fauquier&district-39-piedmont&district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717071;s:14:"count_meetings";i:61;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:130:"https://aablueridge.org/wp-admin/admin-ajax.php?district=district-39-piedmont&district-40-skyline&district-unknown&action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717080;s:14:"count_meetings";i:25;s:4:"name";s:15:"aablueridge.org";s:16:"parent_region_id";i:0;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:73:"https://www.aadistrict30va.org/wp/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763716962;s:14:"count_meetings";i:93;s:4:"name";s:27:"District 30 (Fredricksburg)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:90:"https://sheets.code4recovery.org/storage/1LzVrSE-t-iYMnSmiZtg8449rctjJ8Xn9l-nz9_t4H6k.json";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717027;s:14:"count_meetings";i:64;s:4:"name";s:31:"District 38 (Historic Virginia)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:75:"https://www.tidewaterintergroup.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717037;s:14:"count_meetings";i:356;s:4:"name";s:37:"Tidewater Intergroup (14, 23, 24, 32)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}s:61:"https://aaroanoke.org/wp-admin/admin-ajax.php?action=meetings";a:7:{s:6:"status";s:2:"OK";s:11:"last_import";i:1763717051;s:14:"count_meetings";i:98;s:4:"name";s:21:"Roanoke (District 19)";s:16:"parent_region_id";i:-1;s:13:"change_detect";s:8:"disabled";s:4:"type";s:4:"JSON";}}	
```

And I can refresh them without the duplicates appearing XD

**Comment by SIA-WEBSUP** [comment]
It's great that you were able to rectify this situation with your custom script.

I am curious though, is this because you were unable to 'start from scratch' by deleting the existing data sources and re-adding them one by one?

It would be good to know that the data sources can get into this weird state where they can't be deleted.

**Reply by sblack4** [reply]
Yes, we were unable to delete them from the "Import Data Sources" page. But now we can :D 

I'm not sure how this happened - it coincided with a wordpress and php upgrade which might be related?
